tools/classes/PresenceCompiler.ts

```diff
@@ -0,0 +1,285 @@
+import { config } from "dotenv";
+import { existsSync } from "fs";
+import { rm, writeFile } from "fs/promises";
+import { basename, dirname, resolve } from "path";
+import { fileURLToPath } from "url";
+import webpack from "webpack";
+
+import getFolderLetter from "../util/getFolderLetter.js";
+
+const rootPath = resolve(fileURLToPath(new URL(".", import.meta.url)), "../..");
+
+if (!process.env.GITHUB_ACTIONS) config({ path: resolve(rootPath, ".env") });
+
+export const tsconfig = JSON.stringify({
+	extends: "../../../tsconfig.json",
+});
+
+export default class PresenceCompiler {
+	cwd: string;
+
+	constructor(
+		public options?: {
+			cwd?: string;
+			webpack?: webpack.Configuration;
+			silent?: boolean;
+		}
+	) {
+		this.cwd = options?.cwd ?? rootPath;
+	}
+
+	getPresenceFolder(presence: string) {
+		return resolve(this.cwd, "websites", getFolderLetter(presence), presence);
+	}
+
+	async compilePresence(
+		presence: string | string[],
+		options: {
+			output?: string;
+			transpileOnly?: boolean;
+			noEmit?: boolean;
+		} = {}
+	) {
+		const webpackConfig: webpack.Configuration = {
+			mode: "production",
+			devtool: "inline-source-map",
+			resolve: {
+				extensions: [".ts"],
+			},
+
+			output: options.noEmit
+				? undefined
+				: {
+						iife: false,
+						path: options.output,
+						filename: "[name].js",
+				  },
+			module: {
+				rules: [
+					{
+						test: /\.ts$/,
+						loader: "ts-loader",
+						exclude: /node_modules/,
+						options: {
+							transpileOnly: options.transpileOnly,
+							errorFormatter: (error: any) => {
+								actions.error(chalk.redBright(error.content), {
+									file: error.file,
+									title: `TS ${error.code}`,
+									startLine: error.line,
+								});
+
+								return chalk.cyan(
+									basename(
+										dirname(error.file) +
+											"/" +
+											basename(error.file) +
+											":" +
+											chalk.yellowBright(error.line) +
+											":" +
+											chalk.yellowBright(error.character) +
+											" - " +
+											chalk.redBright("Error ") +
+											chalk.gray("TS" + error.code + ": ") +
+											error.content
+									)
+								);
+							},
+						},
+					},
+				],
+			},
+			...(this.options?.webpack || {}),
+		};
+
+		if (!options.transpileOnly) options.transpileOnly = false;
+
+		if (Array.isArray(presence)) {
+			let errors: webpack.WebpackError[] = [];
+
+			actions.info(chalk.yellow(`Compiling ${presence.length} Presence(s)`));
+			for (const p of presence) {
+				const presencePath = this.getPresenceFolder(p);
+
+				await writeFile(resolve(presencePath, "tsconfig.json"), tsconfig);
+				await this.installPresenceDependencies(p);
+
+				const job = await new Promise<{
+					error: Error | undefined;
+					stats: webpack.Stats | undefined;
+				}>(r =>
+					webpack.webpack(
+						{
+							...webpackConfig,
+							context: presencePath,
+							output: options.noEmit
+								? undefined
```

Comment:
ditto

Reply:
I think you forgot to swap the ternary around here

Reply:
oops